### PR TITLE
feat: generate simple generic helm chart

### PR DESCRIPTION
**Pull request by csviri**
This is the MVP for helm chart generator, it is "bare minimum" that works, will add subsequent PRs to improve.

### CODE REVIEW

File: core/deployment/src/main/java/io/quarkiverse/operatorsdk/deployment/helm/DisableDefaultHelmListener.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package io.quarkiverse.operatorsdk.deployment.helm;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import io.dekorate.WithSession;
+import io.dekorate.kubernetes.config.BaseConfigFluent;
+import io.dekorate.kubernetes.config.Configurator;
+
+public class DisableDefaultHelmListener extends Configurator<BaseConfigFluent<?>> implements WithSession {
```
Comment (metacosm) [inline core/deployment/src/main/java/io/quarkiverse/operatorsdk/deployment/helm/DisableDefaultHelmListener.java:10]: What's the point of this class?

File: samples/mysql-schema/src/main/resources/application.properties [hunk 2]
```diff
@@ -4,4 +4,5 @@ quarkus.native.additional-build-args=--initialize-at-run-time=org.apache.commons
 quarkus.datasource.username=root
 quarkus.datasource.devservices.image-name=mariadb:10.7
 quarkus.kubernetes-client.devservices.override-kubeconfig=true
-quarkus.http.test-port=0
\ No newline at end of file
+quarkus.http.test-port=0
+quarkus.kubernetes.image-pull-policy=IfNotPresent
```
Comment (metacosm) [inline samples/mysql-schema/src/main/resources/application.properties:8]: Why was this added?

Comment (csviri) [reply samples/mysql-schema/src/main/resources/application.properties:8]: I think that was just to test with minikube locally, can remove for now

File: core/deployment/src/main/java/io/quarkiverse/operatorsdk/deployment/helm/DisableDefaultHelmListener.java (hunk 1 above)
Comment (csviri) [reply core/deployment/src/main/java/io/quarkiverse/operatorsdk/deployment/helm/DisableDefaultHelmListener.java:13]: added comment. it is to disable default helm generator

File: samples/mysql-schema/src/main/resources/application.properties (hunk 2 above)
Comment (csviri) [reply samples/mysql-schema/src/main/resources/application.properties]: done

File: core/runtime/src/main/java/io/quarkiverse/operatorsdk/runtime/HelmConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import io.quarkus.runtime.annotations.ConfigGroup;
+import io.quarkus.runtime.annotations.ConfigItem;
+
+@ConfigGroup
+public class HelmConfiguration {
+
+    /**
+     * Can be used to disable helm chart generation.
+     */
+    @ConfigItem(defaultValue = "true")
```
Comment (metacosm) [inline core/runtime/src/main/java/io/quarkiverse/operatorsdk/runtime/HelmConfiguration.java:12]: This should be false by default, imo. We don't want to generate Helm charts on every build for sure.

Comment (csviri) [reply core/runtime/src/main/java/io/quarkiverse/operatorsdk/runtime/HelmConfiguration.java:12]: IDK, I would rather generate everything, users still can opt out, if something is not needed, but they would be aware that it's there. Also would worth to measure the time difference. Will do some naive measurement

Comment (metacosm) [reply core/runtime/src/main/java/io/quarkiverse/operatorsdk/runtime/HelmConfiguration.java:12]: This is more important in the dev mode, so at the very least, helm chart generation should be disabled in dev mode, though I still think it shouldn't be active by default, the same way we don't generate bundles by default either (which is why the support is in a different module). Since helm support is in the core module, it should be deactivated by default. Generating something all the time just to make people aware of it is a bad idea, imo. This should be addressed by documentation.

File: core/deployment/src/main/java/io/quarkiverse/operatorsdk/deployment/helm/HelmChartProcessor.java [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+            "service.yaml",
+            "serviceaccount.yaml"
+    };
+    public static final String CHART_YAML_FILENAME = "Chart.yaml";
+    public static final String VALUES_YAML_FILENAME = "values.yaml";
+    public static final String CRD_DIR = "crds";
+
+    @BuildStep
+    public void handleHelmCharts(
+            BuildProducer<ArtifactResultBuildItem> dummy,
```
Comment (metacosm) [inline core/deployment/src/main/java/io/quarkiverse/operatorsdk/deployment/helm/HelmChartProcessor.java:59]: If this is needed to ensure this steps runs after another one, please add a comment to that effect, otherwise that parameter might get removed at some point, leading to possibly difficult to diagnose issues.

Comment (metacosm) [reply core/deployment/src/main/java/io/quarkiverse/operatorsdk/deployment/helm/HelmChartProcessor.java:59]: Actually, this is a BuildProducer that isn't used so I don't think there's any point in keeping this.

File: core/runtime/src/main/java/io/quarkiverse/operatorsdk/runtime/HelmConfiguration.java (hunk 3 above)
Comment (csviri) [reply core/runtime/src/main/java/io/quarkiverse/operatorsdk/runtime/HelmConfiguration.java:12]: IDK, can do it for now, but my opinion is that if there is no explicit reason why not to generate it it should be there, even for dev mode, so users can see the helm chart if they add an additional dependent resource. 

Why bundle generation is a separate module, I mean why it is not just a flag to not generate it?

File: core/deployment/src/main/java/io/quarkiverse/operatorsdk/deployment/helm/HelmChartProcessor.java (hunk 4 above)
Comment (csviri) [reply core/deployment/src/main/java/io/quarkiverse/operatorsdk/deployment/helm/HelmChartProcessor.java:59]: without it it does not gets executed

Comment (csviri) [reply core/deployment/src/main/java/io/quarkiverse/operatorsdk/deployment/helm/HelmChartProcessor.java:59]: maybe there is a more elegant way to do this

File: core/deployment/src/test/java/io/quarkiverse/operatorsdk/test/HelmChartGeneratorTest.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+public class HelmChartGeneratorTest {
+
+    @RegisterExtension
+    static final QuarkusProdModeTest config = new QuarkusProdModeTest()
+            .setApplicationName("helm-chart-test")
+            .withApplicationRoot(
+                    (jar) -> jar.addClasses(SimpleReconciler.class, SimpleCR.class, SimpleSpec.class, SimpleStatus.class));
+
+    @Test
+    void generatesHelmChart() {
```
Comment (Sgitario) [inline core/deployment/src/test/java/io/quarkiverse/operatorsdk/test/HelmChartGeneratorTest.java:21]: I think it would be good to assert that the resources are indeed generated.

**Review by Sgitario — APPROVED** [review]
I see that you have followed the approach of using templates instead of YAML expressions which is perfectly fine and I think it fits better in your use case. 

I've added a few comments about the tests, but not blockers at all. 

Good job!!

File: core/deployment/src/test/java/io/quarkiverse/operatorsdk/test/HelmChartGeneratorTest.java (hunk 5 above)
Comment (csviri) [reply core/deployment/src/test/java/io/quarkiverse/operatorsdk/test/HelmChartGeneratorTest.java:21]: yes, will take a look on this, there was an issue with the test, and though an e2e will be enough, bit will give it an other try. Thx!

File: samples/exposedapp/src/test/java/io/halkyon/HelmDeploymentE2ETest.java [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+import io.javaoperatorsdk.operator.Operator;
+import io.javaoperatorsdk.operator.api.reconciler.Constants;
+import io.quarkus.test.junit.QuarkusTest;
+
+// currently there is a bug with kind testcontainer that blocks the e2e tests done properly
+// will finish this test when it gets resolved:
+// https://github.com/dajudge/kindcontainer/issues/235
+@Disabled
+@QuarkusTest
+class HelmDeploymentE2ETest {
```
[the remark this answers is not in the document]
Comment (csviri) [reply samples/exposedapp/src/test/java/io/halkyon/HelmDeploymentE2ETest.java:34]: Yeah, it is quite annoying when you need those tools, but managing such tests is quite handy from java code. Will take a look on this in next pr. An other possibility is to commit something like this script to the repo:
https://helm.sh/docs/intro/install/#from-script
And install helm during the test if not installed yet

[the remark this answers is not in the document]
Comment (csviri) [reply samples/exposedapp/src/test/java/io/halkyon/HelmDeploymentE2ETest.java:34]: For GitHub Actions there is a nice action that does that.

File: core/deployment/src/test/java/io/quarkiverse/operatorsdk/test/HelmChartGeneratorTest.java (hunk 5 above)
Comment (csviri) [reply core/deployment/src/test/java/io/quarkiverse/operatorsdk/test/HelmChartGeneratorTest.java:21]: added simple asserts, not sure if we should check the templates in depth in this case.

File: core/runtime/src/main/java/io/quarkiverse/operatorsdk/runtime/HelmConfiguration.java (hunk 3 above)
Comment (metacosm) [reply core/runtime/src/main/java/io/quarkiverse/operatorsdk/runtime/HelmConfiguration.java]: > IDK, can do it for now, but my opinion is that if there is no explicit reason why not to generate it it should be there, even for dev mode, so users can see the helm chart if they add an additional dependent resource.

The core principle is to be as efficient as possible so sure, we can always say things are "fast" enough but from my point of view, if something doesn't need to be done all the time, it shouldn't.

> 
> Why bundle generation is a separate module, I mean why it is not just a flag to not generate it?

Cleaner separation of concerns. Also allows to not bring the dependencies if not used. For that matter, we might want to extract Helm support in its own module at some later point as well.

File: core/deployment/src/main/java/io/quarkiverse/operatorsdk/deployment/helm/HelmChartProcessor.java [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+        });
+
+    }
+
+    @SuppressWarnings("rawtypes")
+    private void addPrimaryClusterRoleBindings(File helmDir, Collection<QuarkusControllerConfiguration> controllerConfigs) {
+        try (InputStream file = Thread.currentThread().getContextClassLoader()
+                .getResourceAsStream(CRD_ROLE_BINDING_TEMPLATE_PATH)) {
+            if (file == null) {
+                throw new IllegalArgumentException("Template file " + CRD_ROLE_BINDING_TEMPLATE_PATH + " doesn't exist");
```
Comment (csviri) [inline core/deployment/src/main/java/io/quarkiverse/operatorsdk/deployment/helm/HelmChartProcessor.java:120]: There is not much benefit to check for null on a static resource.

**Review by metacosm — APPROVED** [review]
